tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/nvptx_backend_lib.cc

```diff
@@ -446,25 +484,112 @@ void GPUBackendInit(const HloModuleConfig& hlo_module_config) {
 
 }  // namespace
 
-StatusOr<string> CompileToPtx(llvm::Module* module,
-                              std::pair<int, int> compute_capability,
+namespace nvptx {
+
+StatusOr<string> CompileToPtx(llvm::Module* module, GpuVersion gpu_version,
                               const HloModuleConfig& hlo_module_config,
                               const string& libdevice_dir_path) {
   static std::once_flag backend_init_flag;
-  std::call_once(backend_init_flag, GPUBackendInit, hlo_module_config);
+  std::call_once(backend_init_flag, NVPTXBackendInit, hlo_module_config);
 
   string ptx;
+  std::unique_ptr<llvm::TargetMachine> target_machine;
   {
     tensorflow::profiler::TraceMe activity(
         [&] { return absl::StrCat("Compiling IR:", module->getName().str()); },
         tensorflow::profiler::TraceMeLevel::kInfo);
     XLA_SCOPED_LOGGING_TIMER("Compile module " + module->getName().str());
-    TF_ASSIGN_OR_RETURN(
-        ptx, CompileModuleToPtx(module, compute_capability, hlo_module_config,
-                                libdevice_dir_path));
+
+    // If the module has no functions or globals, there's nothing to compile.
+    // Just return an empty string.
+    if (module->empty() && module->global_empty()) {
+      VLOG(2) << "Module '" << module->getName().str()
+              << "' is empty. Skipping compilation.";
+      return string();
+    }
+
+    auto compute_capability = absl::get_if<std::pair<int, int>>(&gpu_version);
+    if (compute_capability) {
+      llvm::Triple target_triple("nvptx64-unknown-unknown");
+      // Construct LLVM TargetMachine for NVPTX.
+      std::unique_ptr<llvm::TargetMachine> target_machine =
+          NVPTXGetTargetMachine(target_triple, *compute_capability,
+                                hlo_module_config);
+
+      // Link with libdeivce, and optimize the LLVM module.
+      TF_RETURN_IF_ERROR(LinkAndOptimizeModule(
+          module, gpu_version, hlo_module_config, libdevice_dir_path,
+          NVPTXTargetModuleLinker, target_triple,
+          target_machine.get(), kDefaultInlineThreshold));
+
+      // Lower optimize LLVM module to PTX.
+      TF_ASSIGN_OR_RETURN(ptx, EmitModuleToPTX(module, target_machine.get()));
+    }
```

Comment:
Please return an error if compute_capability was not specified or do not use the generic GpuVersion.

Reply:
return `xla::InternalError` to handle the error.